R/utils.R

```diff
@@ -437,18 +442,184 @@ rot_x <- function(angle=30, hjust = 1, vjust = NULL, vertical = FALSE, ...) {
+#' @param at a character vector of list names to rotate.
+#' @param re a regular expression for selecting names to be used for `at`.
+#' @param axis which axis to rotate.
+#' @param ... additional arguments passed to [rot_x()] or [rot_y()].
+#'
+#' @details
+#' Note that all plots in the list need to be named.
+#'
+#' @seealso [rot_xy()], [rot_x()], [rot_y()].
+#'
+#' @examples
+#' data <- pmplots_data_id()
+#'
+#' co <- c("STUDYc", "CPc", "RF")
+#' etas <- paste0("ETA", 1:3)
+#'
+#' x <- eta_cat(data, x = co, y = etas)
+#' names(x)
+#'
+#' x <- rot_at(x, at = "ETA1vRF", angle = 35)
+#' x$ETA1vRF
+#'
+#' x <- rot_at(x, re = "RF", vertical = TRUE)
+#' x$ETA2vRF
+#' x$ETA3vRF
+#'
+#' @md
+#' @export
+rot_at <- function(x, at = names(x), re = NULL, axis = c("x", "y"), ...) {
+  if(!is.list(x) || is.ggplot(x) || inherits(x, "patchwork")) {
+    abort("`x` must be a list of gg or patchwork objects.")
+  }
+  if(!is_named(x)) abort("`x` must be named.")
+  axis <- match.arg(axis)
+  if(is.character(re)) {
+    where <- grep(re, names(x), perl = TRUE)
+  } else {
+    if(!is.character(at)) abort("`at` must be character.")
+    bad <- setdiff(at, names(x))
+    if(length(bad)) {
+      names(bad) <- rep("x", length(bad))
+      abort("requested names not found in `x`.", body = bad)
+    }
+    where <- which(names(x) %in% at)
+  }
+  if(!length(where)) {
+    warn("did not find any plots for axis rotation.")
+    return(x)
+  }
+  for(w in where) {
+    x[[w]] <- .rotxy(x[[w]], axis = axis, ...)
+  }
+  x
+}
+
+#' Rotate the x or y axes for gg or patchwork objects
+#'
+#' This is a wrapper function around [rot_at()], [rot_x()], and [rot_y()] that
+#' accepts gg objects, patchwork objects, or lists of gg or patchwork objects.
+#' Arguments to those wrapped functions can be passed through. See arguments
+#' to [rot_at()] as well as [rot_x()] and [rot_y()].
+#'
+#' @param x a gg object, patchwork object, or list of those objects.
+#' @param axis which axis to rotate (`x` or `y`).
+#' @param ... arguments passed to [rot_at()] (`at`, `re`) or [rot_x()] / [rot_y()]
+#' (`angle`, `vertical` `hjust`, `vjust`).
+#'
+#' @seealso [rot_at()], [rot_x()], [rot_y()].
+#'
+#' @examples
+#' library(patchwork)
+#'
+#' data <- pmplots_data_obs()
+#' id <- pmplots_data_id()
+#'
+#' p <- dv_pred(data)
+#'
+#' p1 <- rot_xy(p)
+#' p2 <- rot_xy(p, vertical = TRUE)
+#' p3 <- rot_xy(p, axis = "y", angle = 15)
+#'
+#' p1 + p2 + p3
+#'
+#' rot_xy(p + dv_ipred(data))
+#'
+#' co <- c("STUDYc", "CPc")
+#' etas <- paste0("ETA", 1:2)
+#'
+#' x <- eta_covariate(data, x = co, y = etas, transpose = TRUE)
+#'
+#' pp <- rot_xy(x, at = "CPc")
+#'
+#' pm_with(pp, STUDYc/CPc)
+#'
+#' @export
+rot_xy <- function(x, ...) UseMethod("rot_xy")
+
+#' @rdname rot_xy
+#' @export
+rot_xy.gg <- function(x, axis = c("x", "y", ...)) {
+  .rotxy(x, axis = axis, ...)
+}
+
+#' @rdname rot_xy
+#' @export
+rot_xy.patchwork <- function(x, axis = c("x", "y", ...)) {
```

Comment:
The placement of the `...` here leads to the CI failures other than the 3.6 one fixed by gh-97:

https://github.com/metrumresearchgroup/pmplots/actions/runs/9933481761/job/27436292590

```
 ── 1. Error ('test-utils.R:103:3'): rot_xy ─────────────────────────────────────
Error in `.rotxy(x, axis = axis, ...)`: '...' used in an incorrect context
Backtrace:
    ▆
 1. ├─pmplots::rot_xy(g1) at test-utils.R:103:3
 2. └─pmplots:::rot_xy.gg(g1) at pmplots/R/utils.R:603:11
 ```

Another failure surfaces once the `...` is moved outside the axis value:

```
Error (test-utils.R:103:3): rot_xy
Error in `if (axis == "x") {
    x <- x + rot_x(...)
} else {
    x <- x + rot_y(...)
}`: the condition has length > 1
Backtrace:
    ▆
 1. ├─pmplots::rot_xy(g1) at test-utils.R:103:3
 2. └─pmplots:::rot_xy.gg(g1) at pmplots/R/utils.R:603:11
 3.   ├─pmplots:::.rotxy(x, axis = axis, ...) at pmplots/R/utils.R:608:3
 4.   └─pmplots:::.rotxy.gg(x, axis = axis, ...) at pmplots/R/utils.R:477:11
```

Looking quickly, I think that's probably just a matter of a missing `match.arg`.

So perhaps the complete fix for these two failures is something like this

```diff
diff --git a/R/utils.R b/R/utils.R
index b961320..6567287 100755
--- a/R/utils.R
+++ b/R/utils.R
@@ -604,13 +604,15 @@ rot_xy <- function(x, ...) UseMethod("rot_xy")

 #' @rdname rot_xy
 #' @export
-rot_xy.gg <- function(x, axis = c("x", "y", ...)) {
+rot_xy.gg <- function(x, axis = c("x", "y"), ...) {
+  axis <- match.arg(axis)
   .rotxy(x, axis = axis, ...)
 }

 #' @rdname rot_xy
 #' @export
-rot_xy.patchwork <- function(x, axis = c("x", "y", ...)) {
+rot_xy.patchwork <- function(x, axis = c("x", "y"), ...) {
+  axis <- match.arg(axis)
   .rotxy(x, axis = axis, ...)
 }
```

Reply:
Ugh ... just seeing this now too as I was updating the example code. Thanks, yes those `...` should be passed along, not absorbed in to that axis argument. Will push a fix now.

Reply:
Thanks, @kyleam; got this untangled with your suggestion. I believe it's behaving now.